File: afdb/afdb_scrape.R
#!/usr/bin/env Rscript

################################################################################
# afdb/afdb-scrape.R                                                           #
#                                                                              #
# Copyright 2021 Evans Policy Analysis and Research Group (EPAR).              #
#                                                                              #
# This project is licensed under the 3-Clause BSD License. Please see the      #
# license.txt file for more information.                                       #
#                                                                              #
# this script crawls the african development bank project pages to create a    #
# database of specified information                                            #
#                                                                              #
#                                                                              #
# inputs: none                                                                 #
# outputs: a csv file containing project numbers, description, planned         #
#          completion date, dac sector codes, etc                              #
################################################################################

# IMPORTS #
# install.packages("readxl")
# install.packages("rvest")
#install.packages("R.utils",repos = "http://cran.us.r-project.org")
# install.packages("sjmisc")
# install.packages("xml2")
library(readxl)
library(rvest)
library(R.utils)
library(sjmisc)
library(xml2)



# CONSTANTS AND OUTPUT #
DEBUG <- FALSE
RATE_LIMIT <- 10
BASE_URL <- "https://projectsportal.afdb.org/dataportal/VProject/show/"
OUTPUT_FILE <- if(DEBUG) "../data/afdb_test.csv" else "../data/afdb_data.csv"
DAC_FILE <- "../DAC-CRS-CODES.xls"
AFDB_SPREADSHEET <- if(DEBUG) "afdb-ids-short.xlsx" else "afdb-ids.xlsx"
AFDB_SPREADSHEET_URL <- "https://projectsportal.afdb.org/dataportal/VProject/exportProjectList?reportName=dataPortal_project_list"
sink("output.txt", split=TRUE, append = FALSE)

# FUNCTIONS #
#Scrape
get_html <- function(url) {
  tryCatch(
    read_html(url),
    error = function(e) {
      cat("Request failed, retrying.\n")
      Sys.sleep(5) 
      NULL
    }
  )
}

# File write
write_file <- function(df, filename) {
  tryCatch(
    {
      df %>% write.csv( 
        file = filename, 
        quote = TRUE)
      TRUE
    },
    error = function(e) {
      cat("writing to file failed, retrying.\n")
      print(e)
      FALSE
    }
  )
}

#Searches every table in the html page for x1's data
all_table_parse <- function(page, x1) {
  ret <- ""
  tables <- page %>% html_nodes("table")
  #For each table on the webpage
  for(i in 1:length(tables)){
    table <- tryCatch({
      tables %>% .[[i]] %>% html_table(header = FALSE, trim=TRUE)
    }, error = function(err){
      if(DEBUG) print(err)
      NULL
    })
    
    if(is.null(tables[[i]]) || length(tables[[i]]) == 0) next
    
    #Does current table have bad formatting?
    if(str_contains(table[['X1']], "\n\t") && !str_contains(table[['X1']], "Download")){
      #Correct formatting using regex (sorry)
      table <- str_replace_all(table[['X1']], "[^\\S ]+[ ]+[^\\S ]+", "***")
      table <- strsplit(table, "\\*\\*\\*")
      #Search through table rows for matches (i.e. filter)
      if(!is.null(table) && length(table) > 0 && str_contains(table, x1)){
        for(j in 1:length(table)){
          if(str_contains(table[[j]][[1]], x1)){
            ret <- paste(ret, table[[j]][[2]], sep = ", ")
          }  
        }
        #Return if we've found the indicator
        if(length(ret) > 0){
          return(substring(ret, first = 3))
        }
      }
      else{
        next
      }
    }
    
    #Current table did not have bad formatting, just search
    if(!is.null(table) && length(table) > 1){
      for(k in 1:length(table)){
        if(str_contains(table[['X1']][k], x1)){
          return (table[['X2']][k])
        }
      }  
    }
  }
  return(NULL)
}

#Cleaner than all_table_parse, but less generalizable
# (only searches the first table in the HTML)
main_table_parse <- function(page, x1) {
  temp <- (page %>% 
             html_node(".table") %>%
             html_table(header = FALSE) %>%
             filter(X1 == x1))
  # check if table had requested data, if not replace with "N/A"
  return (if(nrow(temp) == 0) "N/A" else temp[['X2']])
}

#Looks up the DAC or DAC5 code in the DAC-CRS-CODES excel file
get_dac5_desc <- function(code){
  if(is.null(code) || code == "N/A") return("N/A")
  code <- strtoi(code)
  desc <- if(code < 1000)
            dac_df$DESCRIPTION[which(dac_df$`DAC 5 CODE` == code)] 
          else
            dac_df$DESCRIPTION[which(dac_df$`concatenate` == code)]
  return(if(!is.null(desc)) desc else "N/A")
}

# MAIN #

#If real run, download the projects file from the AFDB website 
if(!DEBUG) download.file(AFDB_SPREADSHEET_URL, AFDB_SPREADSHEET, method="curl") 
proj_ids <- read_excel(AFDB_SPREADSHEET, skip=0)
#Drop projects that aren't in progress. Must be done in 2 steps
proj_ids <- proj_ids[(proj_ids$Status == "Approved" | proj_ids$Status == "Implementation"), ]
proj_ids <- proj_ids[!is.na(proj_ids$`Project Code`), ]
#Read the DAC description excel file
dac_df <- read_excel(DAC_FILE, sheet=12, skip=2)

# create a dataframe to hold the scraped data.
data <- data.frame(matrix(ncol = 18, nrow = 0))
names(data) <- c(
  "Project ID",
  "Country",
  "Project Title",
  "Description",  
  "Commitment in U.A.",
  "Status",
  "Start Date",
  "Closing Date",
  "Project Duration",
  "Source of Financing",
  "Sovereign",
  "Sector",
  "DAC Sector Code",
  "DAC5 Code",
  "DAC5 Description",
  "Detailed Description",
  "Contact Name",
  "Contact Email"
)


# for each id in provided file
for(id in proj_ids$`Project Code`) {
  if(is.null(id) || is.na(id)) next
  
  start <- Sys.time()
  "scraping project" %>% paste(id, "\n", sep = " ") %>% cat()
  page  <- NULL
  count <- 0
  
  # using a loop to retry on failed attempts just in case there is an http
  # request error. only retry 20 times in case there is a bad id that doesn't
  # work (it happened, don't remove the loop limiter)
  while(length(page) == 0 && count < 20){
    page <- get_html(paste(BASE_URL, id, sep=""))
    count <- count + 1
  }
  # if the loop ended because of the loop limiter skip to next id
  if(length(page) == 0) next
  processing_time <- Sys.time()
  ##############################
  # parse the web page results #
  ##############################
  # pull info from main web page table, only look at active projects
  status <- main_table_parse(page, "Status")
  if(status != "Approved" && status != "Implementation"){
    print(paste("Skipping project:", id, "with status:", status))
    next
  }
  
  commitment <- main_table_parse(page, "Commitment")
  commitment <- if (str_length(commitment) > 5) substring(commitment, 5) else commitment
  approval_date <- main_table_parse(page, "Approval Date")
  completion_date <- main_table_parse(page, "Planned Completion Date")
  duration <- round(difftime(as.Date(completion_date, format="%d %b %Y"),
                             as.Date(approval_date, format="%d %b %Y"),
                             units="days")
                    / 365.25, 2)
  duration <- if(!is.null(duration)) duration else "N/A"
  sov <- main_table_parse(page, "Sovereign / Non-Sovereign")

  sector <- main_table_parse(page, "Sector")
  dac <- main_table_parse(page, "DAC Sector Code")
  if(nchar(dac) < 3) dac <- "N/A"
  dac5 = substring(dac, 1,3)
  dac5_desc <- get_dac5_desc(dac5)
  dac5_desc_detailed <- get_dac5_desc(dac)
  
  contact_name <- all_table_parse(page, "Name")
  contact_name <- if (!is.null(contact_name)) str_to_title(contact_name) else "N/A"
  contact_email <- all_table_parse(page, "Email")
  contact_email <- if(!is.null(contact_email)) contact_email else "N/A"
  funding <- all_table_parse(page, "Funding")
  
  #Get country & project names
  country_plus_project <- page %>% html_node("h2") %>% html_text2()
  country_plus_project <- str_replace(country_plus_project, " ?[\\p{Pd}|:] ?", " - ")
  country_plus_project <- str_split(country_plus_project, " - ", n=2)[[1]]
  country <- if(is.na(country_plus_project[1])) "N/A" else country_plus_project[1]
  project <- if(is.na(country_plus_project[2])) "N/A" else country_plus_project[2]
  #If the country exists and the project doesn't, that usually means that the project name was not of the form "country - project"
  if(country != "N/A" && project == "N/A"){
    project <- country
    country <- "N/A"
  }
  #Reassign country to the (standardized) country name from the table
  country <- all_table_parse(page, "Country")
  
  # by default set to n/a will replace later if we find what we're after
  desc <- ""
  # pull out all the divs and put them in a vector like object
  divs <- page %>% html_nodes("div")
  # iterate through the list looking for information
  for(div in divs) {
    if(html_text(html_node(div,"h3")) %in% "Project General Description"){
      desc <- div %>% html_node("p") %>% html_text()
    }
    if(html_text(html_node(div,"h3")) %in% "Project Objectives") {
      desc <- paste(div %>% html_node("p") %>% html_text(), desc, sep="\n\n")
    }
  }
  
  if(DEBUG) print(paste(id, country, project, commitment, status, approval_date, completion_date, duration, funding, sov, sector, dac, dac5, dac5_desc, dac5_desc_detailed, contact_name, contact_email, sep="; "))
  # add parsed data into the main data frame
  data[nrow(data) + 1,] <- c(id, country, project, desc, commitment, status, approval_date, completion_date, duration, funding, sov, sector, dac, dac5, dac5_desc, dac5_desc_detailed, contact_name, contact_email)

  # calculate elapsed time for request
  elapsed <- Sys.time() - start
  # Respect the robots.txt - only do a request every 10 seconds
  printf("Download time: %.2f // Processing time: %.2fs // Sleep time: %.2fs\n", processing_time - start, elapsed - processing_time + start, RATE_LIMIT - elapsed)
  if(elapsed < RATE_LIMIT) {
    Sys.sleep(RATE_LIMIT - elapsed)
  }
}

# Write to disk
written <- FALSE
count <- 0
while(!written && count < 20){
	written <- write_file(data, OUTPUT_FILE)
	count <- count + 1
}

closeAllConnections()
print("All done!")

File: afdb/afdb_scrape.py
################################################################################
# afdb/afdb-scrape.py                                                          #
#                                                                              #
# Copyright 2021 Evans Policy Analysis and Research Group (EPAR).              #
#                                                                              #
# This project is licensed under the 3-Clause BSD License. Please see the      #
# license.txt file for more information.                                       #
#                                                                              #
# This script crawls the African Development Bank project pages to             #
# create a database of specified information                                   #
################################################################################

__copyright__ = """
Copyright 2021 Evans Policy Analysis and Research Group (EPAR).
"""
__license__ = """
This project is licensed under the 3-Clause BSD License. Please see the 
license.txt file for more information.
"""

# Imports
import html
import pandas as pd
import requests
import sys
import time
from bs4 import BeautifulSoup

# Constants
DEBUG = False if len(sys.argv) == 1 else sys.argv[1] == "True"
DEBUG_COUNT = 5 # How many projects to scrape when debugging

BASE_URL = 'https://projectsportal.afdb.org/dataportal/VProject/show/'
CWD = './data/'
PROJECT_LIST_URL = 'https://projectsportal.afdb.org/dataportal/VProject/exportProjectList?reportName=dataPortal_project_list'
PROJECT_LIST = CWD + 'afdb_ids_debug.xlsx' if DEBUG else CWD + 'afdb_ids.xlsx'
OUTPUT_FILE = CWD + 'afdb_data_debug.xlsx' if DEBUG else CWD + 'afdb_data.xlsx'
SCRAPE_DELAY_IN_SEC = 5
UA_TO_USD_MULTIPLIER = 1.39589
# Key = AfDB country name format, Value = IFI project country name format
IFI_COUNTRIES = { 
    'Angola': 'Angola',
    'Benin' : 'Benin',
    'Botswana' : 'Botswana',
    'Burkina Faso' : 'Burkina Faso',
    'Burundi' : 'Burundi',
    'Cameroon' : 'Cameroon',
    'Cape Verde' : 'Cabo Verde',
    'Central African Republic' : 'Central African Republic',
    'Chad' : 'Chad',
    'Comoros' : 'Comoros',
    "Cote d'Ivoire" : "Côte d'Ivoire",
    'Congo, the Democratic Republic of the' : 'Democratic Republic of the Congo',
    'Equatorial Guinea' : 'Equatorial Guinea',
    'Eritrea' : 'Eritrea',
    'Eswatini' : 'Eswatini',
    'Ethiopia' : 'Ethiopia',
    'Gabon' : 'Gabon',
    'Gambia' : 'Gambia',
    'Ghana' :  'Ghana',
    'Guinea' : 'Guinea',
    'Guinea-Bissau' : 'Guinea-Bissau',
    'Kenya' : 'Kenya',
    'Kingdom of Lesotho' : 'Lesotho',
    'Liberia' : 'Liberia',
    'Madagascar' : 'Madagascar',
    'Malawi' : 'Malawi',
    'Mali' : 'Mali',
    'Mauritania' : 'Mauritania',
    'Mauritius' : 'Mauritius',
    'Mozambique' : 'Mozambique',
    'Namibia' : 'Namibia',
    'Niger' : 'Niger',
    'Nigeria' : 'Nigeria',
    'Congo' : 'Republic of the Congo',
    'Rwanda' : 'Rwanda',
    'Sao Tome and Pricipe' : 'Sao Tome and Principe',
    'Senegal' : 'Senegal',
    'Seychelles' : 'Seychelles',
    'Sierra Leone' : 'Sierra Leone',
    'South Africa' : 'South Africa',
    'South Sudan' : 'South Sudan',
    'Tanzania, United Republic of' : 'Tanzania',
    'Togo' : 'Togo',
    'Uganda' : 'Uganda',
    'Zambia' : 'Zambia',
    'Zimbabwe' : 'Zimbabwe',
    #Include regions
    'Multinational' : 'Multinational'
}

try:
    DAC_LOOKUP = pd.read_excel('./DAC-CRS-CODES.xls', sheet_name='Purpose codes', header=2)
except Exception as e:
    print("Exception opening DAC code excel file: {0}".format(e))
    print("This error usually happens when running from the script from the wrong directory. Make sure to run from '411-IFI-Aid/'")

# Downloads the current list of AfDB projects
def download_afdb_projects_list():
    print('Downloading projects spreadsheet from the AfDB website')
    r = requests.get(PROJECT_LIST_URL)
    print('Download complete!')
    unfiltered_projs = open(PROJECT_LIST, 'wb')
    unfiltered_projs.write(r.content)
    unfiltered_projs.close()
    print('Writing unfiltered projects to ' + PROJECT_LIST)

# Download url and return a BeautifulSoup object from the resulting html
def get_html(url):
    attempts = 0
    # Retry up to 20 times, then quit
    while(attempts < 20):
        try:
            attempts += 1
            r = requests.get(url)
            clean_html = html.unescape(r.text)
            clean_html = "".join(line.strip() for line in clean_html.split("\n"))
            return BeautifulSoup(clean_html, 'html.parser')
        except (Exception) as e:
            print('Failed to download webpage, trying again')
            print(e)
            time.sleep(5)
    return None

# Find and return data from standard tables on the project page
def find_in_table(soup, var):
    try:
        temp = soup.body.find(text=var).parent.parent.find_next('td').contents[0]
        return temp
    except (Exception) as e:
        return ""

# Find and return data from nonstandard tables on the project page
def find_in_nonstandard_table(soup, var):
    try:
        temp = soup.body.find(text=var).find_parent(class_='col-md-4').find_next(class_='col-md-8').contents[0]
        return temp
    except (Exception) as e:
        return ""

# Find and return data from a project page's heading
def find_in_heading(soup, title):
    try: 
        temp = soup.body.find(text=title).parent.find_next('p').contents[0]
        return temp
    except (Exception) as e:
        return ""

# Returns the description of the given DAC code from the local DAC code spreadsheet
def get_dac5_desc(code):
    if code == None or len(code) < 3 or code == 'N/A':
        return "N/A"
    code = int(code)
    column_name = 'DAC 5 CODE' if code < 1000 else 'concatenate'
    desc = DAC_LOOKUP[DAC_LOOKUP[column_name] == code]['DESCRIPTION'].values[0]
    return desc

# Main
if not DEBUG:
    download_afdb_projects_list()

# Read in the unfiltered list of projects
df = pd.read_excel(PROJECT_LIST)
print('Filtering to active projects in IFI countries')
project_ids = df[(df['Status'].isin(['Approved', 'Implementation']) & df['Country'].isin(IFI_COUNTRIES.keys()))]

scraped_data = []
count = 0
# Scrape each project and store in scraped_data
for index, row in project_ids.iterrows():
    # Drop projects that are not in countries of interest, approved, or in progress
    if row['Project Code'] == None or row['Project Code'] == '' or (row['Status'] != 'Approved' and row['Status'] != 'Implementation'):
        next

    # Start timer
    start_time = time.time()
    
    print('\n\nScraping project: ' + row['Project Code'])
    data = {}
    soup = get_html(BASE_URL + row['Project Code'])

    # Get details from html
    data['IFI'] = 'African Development Bank'
    data['Project ID'] = row['Project Code']
    country = find_in_table(soup, 'Country')
    data['Country'] = IFI_COUNTRIES[country.get_text() if type(country) != str else country]
    # Break down the "Country - Project Title" header to get the title
    cpt = soup.body.find('h2', class_='title').get_text().split('- ', 1)
    title = cpt[1] if len(cpt) == 2 else cpt[0]
    data['Project Title'] = title
    data['Status'] = find_in_table(soup, 'Status')
    ua_commitment = find_in_table(soup, 'Commitment').split(' ', 1)[1]
    data['Commitment Amount (USD)'] = int(float(ua_commitment.replace(',','')) * UA_TO_USD_MULTIPLIER) if ua_commitment != None else None
    #data['Source of Financing'] = find_in_nonstandard_table(soup, 'Funding').get_text()
    #data['Sovereign'] = find_in_table(soup, 'Sovereign / Non-Sovereign').get_text()
    start_date = pd.to_datetime(find_in_table(soup, 'Approval Date'), infer_datetime_format=True)
    closing_date = pd.to_datetime(find_in_table(soup, 'Planned Completion Date'), infer_datetime_format=True)
    data['Project Duration'] = round((closing_date - start_date).days / 365.25, 2)
    data['Approval Date'] = start_date.date().isoformat()
    data['Closing Date'] = closing_date.date().isoformat()
    data['Description'] = str(find_in_heading(soup, 'Project General Description')) 
    obj = str(find_in_heading(soup, 'Project Objectives'))
    data['Description'] += "\n" + obj if (obj != "" or obj == None) else ""
    data['Project Contact'] = str(find_in_table(soup, 'Name')).title()
    data['Contact Details'] = find_in_table(soup, 'Email')
    data['DAC Sector Code'] = find_in_table(soup, 'DAC Sector Code')
    data['Detailed Description'] = get_dac5_desc(data['DAC Sector Code'])
    data['DAC5 Code'] = data['DAC Sector Code'][:3]
    data['DAC5 Description'] = get_dac5_desc(data['DAC5 Code'])
    data['Primary Sector'] = find_in_table(soup, 'Sector').get_text()
    data['Additional Sectors'] = "{0}; {1}".format(data['DAC5 Description'], data['Detailed Description'])

    # Remove intermediate DAC sector codes to standardize columns across IFIs
    data.pop('DAC Sector Code')
    data.pop('DAC5 Code')
    data.pop('DAC5 Description')
    data.pop('Detailed Description')

    # Print and store scraped project
    [print(key,':',value) for key, value in data.items()]
    scraped_data.append(data)
    
    # Break if debugging and reached limit
    count = count + 1
    if DEBUG and count == DEBUG_COUNT:
        print("Scraped first {0} projects for debugging, ending now".format(DEBUG_COUNT))
        break

    # Make sure to wait before scraping each page (10s delay requested by AfDB's robots.txt)
    elapsed = time.time() - start_time
    if elapsed < SCRAPE_DELAY_IN_SEC:
        print("Waiting %.2fs before next scrape" % (10 - SCRAPE_DELAY_IN_SEC))
        time.sleep(SCRAPE_DELAY_IN_SEC - elapsed)

# Convert into an excel file
print("Creating excel file '%s' with scraped data" % OUTPUT_FILE)
df = pd.DataFrame.from_records(scraped_data)

# Don't fail because the output file was open
while True:
    try:
        df.to_excel(OUTPUT_FILE, index=False, na_rep='', float_format='%.2f')
        break
    except Exception as e:
        print("Failed to write to CSV file. Please make sure that 1) file is closed, and 2) you are running this script from the 411-IFI-Aid/ folder.")
        time.sleep(5)

print('All done!')